ntoskrnl/io/pnpmgr/devnode.c

```diff
@@ -68,10 +70,82 @@ PipAllocateDeviceNode(
         PhysicalDeviceObject->Flags &= ~DO_DEVICE_INITIALIZING;
     }
 
+    DPRINT("Allocated devnode 0x%p\n", DeviceNode);
+
     /* Return the node */
     return DeviceNode;
 }
 
+VOID
+PipInsertDevNode(
+    _In_ PDEVICE_NODE DeviceNode,
+    _In_ PDEVICE_NODE ParentNode)
+{
+    KIRQL oldIrql;
+
+    ASSERT(DeviceNode->Parent == NULL);
+
+    KeAcquireSpinLock(&IopDeviceTreeLock, &oldIrql);
+    DeviceNode->Parent = ParentNode;
+    DeviceNode->Sibling = NULL;
+    if (ParentNode->LastChild == NULL)
+    {
+        ParentNode->Child = DeviceNode;
+        ParentNode->LastChild = DeviceNode;
+    }
+    else
+    {
+        ParentNode->LastChild->Sibling = DeviceNode;
+        ParentNode->LastChild = DeviceNode;
+    }
+    KeReleaseSpinLock(&IopDeviceTreeLock, oldIrql);
+    DeviceNode->Level = ParentNode->Level + 1;
+
+    DPRINT("Inserted devnode 0x%p to parent 0x%p\n", DeviceNode, ParentNode);
+}
+
+PNP_DEVNODE_STATE
+PiSetDevNodeState(
+    _In_ PDEVICE_NODE DeviceNode,
+    _In_ PNP_DEVNODE_STATE NewState)
+{
+    KIRQL oldIrql;
+
+    KeAcquireSpinLock(&IopDeviceTreeLock, &oldIrql);
+
+    PNP_DEVNODE_STATE prevState = DeviceNode->State;
+    if (DeviceNode->State != NewState)
+    {
+        DeviceNode->State = NewState;
+        DeviceNode->PreviousState = prevState;
+        DeviceNode->StateHistory[DeviceNode->StateHistoryEntry++] = prevState;
+        DeviceNode->StateHistoryEntry %= DEVNODE_HISTORY_SIZE;
+    }
+
+    KeReleaseSpinLock(&IopDeviceTreeLock, oldIrql);
+
+    DPRINT1("%wZ Changed state 0x%x => 0x%x\n", &DeviceNode->InstancePath, prevState, NewState);
+
+    return prevState;
+}
+
+VOID
+PiSetDevNodeProblem(
+    _In_ PDEVICE_NODE DeviceNode,
+    _In_ UINT32 Problem)
+{
+    DeviceNode->Flags |= DNF_HAS_PROBLEM;
+    DeviceNode->Problem = Problem;
+}
+
+VOID
+PiClearDevNodeProblem(
+    _In_ PDEVICE_NODE DeviceNode)
+{
+    DeviceNode->Flags &= ~DNF_HAS_PROBLEM;
+    DeviceNode->Problem = 0;
```

Comment:
Unless you will add extra code in these functions, they look like to be macro-ifiable, or inlinable.